splay/src/main/scala/org/scalajs/benchmark/splay/Splay.scala

```diff
@@ -0,0 +1,439 @@
+ *
+ * @constructor
+ */
+class SplayTree[T <: AnyRef] {
+
+  /**
+   * Pointer to the root node of the tree.
+   *
+   * @type {SplayTree.Node}
+   * @private
+   */
+  private var root: Node = null
+
+  /**
+   * @return {boolean} Whether the tree is empty.
+   */
+  @inline
+  def isEmpty: Boolean = root == null
+
+  /**
+   * Inserts a node into the tree with the specified key and value if
+   * the tree does not already contain a node with the specified key. If
+   * the value is inserted, it becomes the root of the tree.
+   *
+   * @param {number} key Key to insert into the tree.
+   * @param {*} value Value to insert into the tree.
+   */
+  def insert(key: Int, value: T): Unit = {
+    if (isEmpty) {
+      root = new Node(key, value)
+    } else {
+      // Splay on the key to move the last node on the search path for
+      // the key to the root of the tree.
+      splay(key)
+      if (root.key != key) {
+        val node = new Node(key, value)
+        if (key > root.key) {
+          node.left = root
+          node.right = root.right
+          root.right = null
+        } else {
+          node.right = root
+          node.left = root.left
+          root.left = null
+        }
+        root = node
+      }
+    }
+  }
+
+  /**
+   * Removes a node with the specified key from the tree if the tree
+   * contains a node with this key. The removed node is returned. If the
+   * key is not found, an exception is thrown.
+   *
+   * @param {number} key Key to find and remove from the tree.
+   * @return {SplayTree.Node} The removed node.
+   */
+  def remove(key: Int): Node = {
+    if (isEmpty) {
+      throw new Exception(s"Key not found: $key")
+    } else {
+      splay(key)
+      if (root.key != key) {
+        throw new Exception(s"Key not found: $key")
+      } else {
+        val removed = root
+        if (root.left == null) {
+          root = root.right
+        } else {
+          val right = root.right
+          root = root.left
+          // Splay to make sure that the new root has an empty right child.
+          splay(key)
+          // Insert the original right child as the right child of the new
+          // root.
+          root.right = right
+        }
+        removed
+      }
+    }
+  }
+
+  /**
+   * Returns the node having the specified key or null if the tree doesn't contain
+   * a node with the specified key.
+   *
+   * @param {number} key Key to find in the tree.
+   * @return {SplayTree.Node} Node having the specified key.
+   */
+  def find(key: Int): Node = {
+    if (isEmpty) {
+      null
+    } else {
+      splay(key)
+      if (root.key == key) root
+      else null
+    }
+  }
+
+  /**
+   * @return {SplayTree.Node} Node having the maximum key value.
+   */
+  def findMax(opt_startNode: Node): Node = {
+    if (isEmpty) {
+      null
+    } else {
+      var current =
+        if (opt_startNode != null) opt_startNode
+        else root
```

Comment:
The idiomatic way to do this in Scala would be to use overloads:
```scala
def findMax(): Node = findMax(root)

def findMax(startNode: Node): Node = {
  ...
  var current = startNode
  ...
}
```
and, it would be more efficient.